f3/powertable.go

```diff
@@ -1,22 +1,75 @@
 package f3
 
-// A power table maps participant IDs to power values.
+import (
+	"sort"
+)
+
+// PowerEntry represents a single entry in the PowerTable.
+// It includes an ActorID and its Weight
+type PowerEntry struct {
+	ID     ActorID
+	Power  *StoragePower
+	PubKey []byte
+}
+
+// PowerTable maps ActorID to a unique index in the range [0, len(powerTable.Entries)).
+// Entries is the reverse mapping to a PowerEntry.
 type PowerTable struct {
-	Entries map[ActorID]uint
-	Total   uint
+	Entries []PowerEntry    // Slice to maintain the order. Meant to be maintained in order in order by (Power descending, ID ascending)
+	Lookup  map[ActorID]int // Maps ActorID to the index of the associated entry in Entries
+	Total   *StoragePower
 }
 
-func NewPowerTable() PowerTable {
-	return PowerTable{
-		Entries: map[ActorID]uint{},
-		Total:   0,
+// NewPowerTable creates a new PowerTable from a slice of PowerEntry .
+// It is more efficient than Add, as it only needs to sort the entries once.
+func NewPowerTable(entries []PowerEntry) *PowerTable {
+	sort.Slice(entries, func(i, j int) bool {
+		return entries[i].Power.Cmp(entries[j].Power) > 0 ||
+			(entries[i].Power.Cmp(entries[j].Power) == 0 && entries[i].ID < entries[j].ID)
+	})
+
+	lookup := make(map[ActorID]int, len(entries))
+	total := NewStoragePower(0)
+	for i, entry := range entries {
+		lookup[entry.ID] = i
+		total.Add(total, entry.Power)
+	}
+
+	return &PowerTable{
+		Entries: entries,
+		Lookup:  lookup,
+		Total:   total,
 	}
 }
 
-func (p *PowerTable) Add(id ActorID, power uint) {
-	if p.Entries[id] != 0 {
+// Add adds a new entry to the PowerTable.
+// This is linear in the table size, so adding many entries this way is inefficient.
+func (p *PowerTable) Add(id ActorID, power *StoragePower, pubKey []byte) {
+	entry := PowerEntry{ID: id, Power: power, PubKey: pubKey}
+	index := sort.Search(len(p.Entries), func(i int) bool {
+		return p.Entries[i].Power.Cmp(power) > 0 ||
+			(p.Entries[i].Power.Cmp(power) == 0 && p.Entries[i].ID < id)
+	})
+
+	// Check for duplication at the found index or adjacent entries
+	if index < len(p.Entries) && (p.Entries[index].ID == id || (index > 0 && p.Entries[index-1].ID == id)) {
```

Comment:
You only need to check the exact entry pointed to.

> Search uses binary search to find and return the **smallest** index i in [0, n) at which f(i) is true

Reply:
Done.